url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/JvmUtil.java

```diff
@@ -0,0 +1,2125 @@
+    public static final int OBJECT_HEADER_SIZE_64_BIT = 12;
+
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_OBJECT_FOR_32_BIT = 4;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_OBJECT_FOR_64_BIT = 8;
+
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_32_BIT_FOR_JAVA_6 = 8;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_64_BIT_WITH_COMPRESSED_REF_FOR_JAVA_6 = 12;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_64_BIT_WITHOUT_COMPRESSED_REF_FOR_JAVA_6 = 16;
+
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_32_BIT_FOR_JAVA_7 = 80;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_64_BIT_WITH_COMPRESSED_REF_FOR_JAVA_7 = 84;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_64_BIT_WITHOUT_COMPRESSED_REF_FOR_JAVA_7 = 160;
+
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_32_BIT_FOR_JAVA_8 = 64;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_64_BIT_WITH_COMPRESSED_REF_FOR_JAVA_8 = 64;
+    public static final int CLASS_DEF_POINTER_OFFSET_IN_CLASS_64_BIT_WITHOUT_COMPRESSED_REF_FOR_JAVA_8 = 120;
+
+    public static final int SIZE_FIELD_OFFSET_IN_CLASS_32_BIT = 12;
+    public static final int SIZE_FIELD_OFFSET_IN_CLASS_64_BIT = 24;
+
+    private static final long TWENTY_FIVE_GB = 25L * 1024L * 1024L * 1024L;
+    private static final long FIFTY_SEVEN_GB = 57L * 1024L * 1024L * 1024L;
+
+    public static final int BOOLEAN_SIZE = 1;
+    public static final int BYTE_SIZE = Byte.SIZE / BYTE;
+    public static final int CHAR_SIZE = Character.SIZE / BYTE;
+    public static final int SHORT_SIZE = Short.SIZE / BYTE;
+    public static final int INT_SIZE = Integer.SIZE / BYTE;
+    public static final int FLOAT_SIZE = Float.SIZE / BYTE;
+    public static final int LONG_SIZE = Long.SIZE / BYTE;
+    public static final int DOUBLE_SIZE = Double.SIZE / BYTE;
+
+    private static final Logger logger = Logger.getLogger(JvmUtil.class);
+
+    private static VMOptions options;
+    private static Unsafe unsafe;
+    private static Object[] objArray;
+    private static int addressSize;
+    private static int headerSize;
+    private static int arrayHeaderSize;
+    private static long baseOffset;
+    private static int indexScale;
+    private static int classDefPointerOffsetInObject;
+    private static int classDefPointerOffsetInClass;
+    private static int sizeFieldOffsetOffsetInClass;
+    private static Integer processId;
+    private static JvmAwareUtil jvmAwareUtil;
+
+    private static final Map<Class<?>, ClassInfo> classCache = new HashMap<Class<?>, ClassInfo>();
+    private static final Map<Class<?>, Map<String, Field>> classFieldCache = new HashMap<Class<?>, Map<String, Field>>();
+    private static final Map<Class<?>, Map<Field, Long>> classFieldOffsetCache = new HashMap<Class<?>, Map<Field, Long>>();
+
+    static {
+        init();
+    }
+
+    private static void init() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 167-198 into a separate method.